rbd: implement GetSnapshot CSI procedure

nixpanic opened the pull request:
A new `GetSnapshot` CSI procedure is introduced.

Depends-on: container-storage-interface/spec#586
See-also: kubernetes/enhancements#5013

**Note:**
This currently uses unmerged changes to the CSI specification, hence this PR is a **Draft**.

Rook is getting prepared for the CSI sidecar updates to prevent panics too (rook/rook#15878).

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic committed:
rbd: implement GetSnapshot CSI procedure

See-also: container-storage-interface/spec#586
Signed-off-by: Niels de Vos <ndevos@ibm.com>

black-dragon74 commented on internal/rbd/controllerserver.go:
```diff
@@ -1701,3 +1701,27 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+	}
+	defer snapshot.Destroy(ctx)
+
+	csiSnap, err := snapshot.ToCSI(ctx)
```
`ToCSI` will fail due to missing `SourceVolumeID` on the fetched snapshot. However, we can get the same from the incoming request.

nixpanic replied:
Indeed, that is interesting! `genSnapFromSnapID()` should return a complete `rbdSnapshot` struct, but this detail is missing. I wonder if it is intentionally not stored/fetched in the journal...

Will look into this a little more soon.

nixpanic replied:
The incoming request does not contain the SourceVolumeId, as that should be the ID (handle) of the parent volue that was used to create the snapshot from.

The 2nd commit adds the `parent.VolID` to the snapshot. Basic testing suggests that everything still works with that.

Now I wonder how/if `VolumeSnapshotContent.spec.source.volumeHandle` was set before this PR?

nixpanic committed:
rbd: fetch the SourceVolumeId (parent) of a snapshot

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

nixpanic commented:
> /test ci/centos/mini-e2e/k8s-1.32

The resizes sidecar does not seem to like the capability (it won't know about it):
```
  I0515 16:24:34.336915 80385 log.go:56] Logs of cephcsi-e2e-bf9a0dd6/csi-rbdplugin-provisioner-84c67cdf4-mszb2:csi-resizer on node minikube

  I0515 16:24:34.336948 80385 log.go:57] STARTLOG

  I0515 16:20:18.786218       1 main.go:111] "Version" version="v1.13.1"
  I0515 16:20:18.787054       1 feature_gate.go:387] feature gates: {map[RecoverVolumeExpansionFailure:true]}
  I0515 16:20:18.787221       1 envvar.go:172] "Feature gate default state" feature="ClientsAllowCBOR" enabled=false
  I0515 16:20:18.787240       1 envvar.go:172] "Feature gate default state" feature="ClientsPreferCBOR" enabled=false
  I0515 16:20:18.787245       1 envvar.go:172] "Feature gate default state" feature="InformerResourceVersion" enabled=false
  I0515 16:20:18.787249       1 envvar.go:172] "Feature gate default state" feature="WatchListClient" enabled=false
  I0515 16:20:18.788215       1 connection.go:234] "Connecting" address="unix:///csi/csi-provisioner.sock"
  I0515 16:20:18.789799       1 common.go:143] "Probing CSI driver for readiness"
  I0515 16:20:18.789842       1 connection.go:264] "GRPC call" method="/csi.v1.Identity/Probe" request="{}"
  I0515 16:20:18.791060       1 connection.go:270] "GRPC response" response="{}" err=null
  I0515 16:20:18.791083       1 connection.go:264] "GRPC call" method="/csi.v1.Identity/GetPluginInfo" request="{}"
  I0515 16:20:18.792150       1 connection.go:270] "GRPC response" response="{\"name\":\"rbd.csi.ceph.com\",\"vendor_version\":\"canary\"}" err=null
  I0515 16:20:18.792173       1 main.go:166] "CSI driver name" driverName="rbd.csi.ceph.com"
  I0515 16:20:18.792183       1 connection.go:264] "GRPC call" method="/csi.v1.Identity/GetPluginCapabilities" request="{}"
  I0515 16:20:18.792798       1 connection.go:270] "GRPC response" response="{\"capabilities\":[{\"service\":{\"type\":\"CONTROLLER_SERVICE\"}},{\"volume_expansion\":{\"type\":\"ONLINE\"}},{\"service\":{\"type\":\"VOLUME_ACCESSIBILITY_CONSTRAINTS\"}}]}" err=null
  I0515 16:20:18.792822       1 connection.go:264] "GRPC call" method="/csi.v1.Controller/ControllerGetCapabilities" request="{}"
  panic: runtime error: invalid memory address or nil pointer dereference
  [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xa97ae2]

  goroutine 1 [running]:
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripSingleValue({0x24249c8, 0xc0000fefc0}, {{}, 0x1c60a00?, 0x0?, 0x1c5c140?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:71 +0x82
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripValue({0x24249c8, 0xc0000fefc0}, {{}, 0x1c60a00?, 0x0?, 0xc00017a5c8?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:94 +0x1b8
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripMessage.func1({0x24249c8, 0xc0000fefc0}, {{}, 0x1c60a00?, 0x0?, 0xc00025d770?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:108 +0xdc
  google.golang.org/protobuf/internal/impl.(*messageState).Range(0xc00017a5a0, 0xc00025d770)
  	/workspace/vendor/google.golang.org/protobuf/internal/impl/message_reflect_gen.go:51 +0x1e3
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripMessage({0x2415050, 0xc00017a5a0})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:103 +0x6d
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripSingleValue({0x24249c8, 0xc0000ff0e0}, {{}, 0x2074fa0?, 0xc00017a5a0?, 0x1c5c140?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:69 +0xd4
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripValue({0x24249c8, 0xc0000ff0e0}, {{}, 0x2074fa0?, 0xc00017a5a0?, 0xc0004bb680?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:94 +0x1b8
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripMessage.func1({0x24249c8, 0xc0000ff0e0}, {{}, 0x2074fa0?, 0xc00017a5a0?, 0xc00025d760?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:108 +0xdc
  google.golang.org/protobuf/internal/impl.(*messageState).Range(0xc000120040, 0xc00025d760)
  	/workspace/vendor/google.golang.org/protobuf/internal/impl/message_reflect_gen.go:58 +0x130
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripMessage({0x2415050, 0xc000120040})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:103 +0x6d
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripSingleValue({0x24249c8, 0xc0000fed80}, {{}, 0x2074fa0?, 0xc000120040?, 0x1c5c140?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:69 +0xd4
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripValue({0x24249c8, 0xc0000fed80}, {{}, 0x1f7d040?, 0xc00017a690?, 0xc00025e700?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:82 +0x20a
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripMessage.func1({0x24249c8, 0xc0000fed80}, {{}, 0x1f7d040?, 0xc00017a690?, 0xc00025d670?})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:108 +0xdc
  google.golang.org/protobuf/internal/impl.(*messageState).Range(0xc00026ed40, 0xc00025d670)
  	/workspace/vendor/google.golang.org/protobuf/internal/impl/message_reflect_gen.go:51 +0x1e3
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripMessage({0x2415050, 0xc00026ed40})
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:103 +0x6d
  github.com/kubernetes-csi/csi-lib-utils/protosanitizer.(*stripSecrets).String(0xc0004bba10)
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:56 +0x4d
  github.com/kubernetes-csi/csi-lib-utils/connection.LogGRPC({0x23fbf38, 0xc00029ea10}, {0x2132464, 0x2c}, {0x1ea2f20, 0xc000495320}, {0x1ed10a0, 0xc00026ed40}, 0xc000334808, 0xc000495350, ...)
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/connection/connection.go:266 +0x290
  google.golang.org/grpc.NewClient.chainUnaryClientInterceptors.func1({0x23fbf38, 0xc00029ea10}, {0x2132464, 0x2c}, {0x1ea2f20, 0xc000495320}, {0x1ed10a0, 0xc00026ed40}, 0xc000334808, 0x220af80, ...)
  	/workspace/vendor/google.golang.org/grpc/clientconn.go:458 +0x118
  google.golang.org/grpc.(*ClientConn).Invoke(0xc000334808, {0x23fbf38?, 0xc00029ea10?}, {0x2132464?, 0xc000495320?}, {0x1ea2f20?, 0xc000495320?}, {0x1ed10a0?, 0xc00026ed40?}, {0x0, ...})
  	/workspace/vendor/google.golang.org/grpc/call.go:35 +0x205
  github.com/container-storage-interface/spec/lib/go/csi.(*controllerClient).ControllerGetCapabilities(0xc0004bbc38, {0x23fbf38, 0xc00029ea10}, 0xc000495320, {0x0, 0x0, 0x0})
  	/workspace/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi_grpc.pb.go:297 +0xc8
  github.com/kubernetes-csi/csi-lib-utils/rpc.GetControllerCapabilities({0x23fbf38, 0xc00029ea10}, 0x35da6a0?)
  	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/rpc/common.go:87 +0x6a
  github.com/kubernetes-csi/external-resizer/pkg/csi.(*client).SupportsControllerResize(0x23fbe58?, {0x23fbf38?, 0xc00029ea10?})
  	/workspace/pkg/csi/client.go:106 +0x2d
  github.com/kubernetes-csi/external-resizer/pkg/resizer.supportsControllerResize({0x24083e0, 0xc0000d0db0}, 0xc000320780?)
  	/workspace/pkg/resizer/csi_resizer.go:250 +0x67
  github.com/kubernetes-csi/external-resizer/pkg/resizer.NewResizerFromClient({0x24083e0, 0xc0000d0db0}, 0x22ecb25c00, {0x2426e88, 0xc000104c40}, {0xc000240260, 0x10})
  	/workspace/pkg/resizer/csi_resizer.go:59 +0xb4
  main.main()
  	/workspace/cmd/csi-resizer/main.go:180 +0x9e5
```

nixpanic commented:
> The resizes sidecar does not seem to like the capability (it won't know about it):
> 
> ```
>   I0515 16:24:34.336915 80385 log.go:56] Logs of cephcsi-e2e-bf9a0dd6/csi-rbdplugin-provisioner-84c67cdf4-mszb2:csi-resizer on node minikube
> 
>   I0515 16:24:34.336948 80385 log.go:57] STARTLOG
> 
>   I0515 16:20:18.786218       1 main.go:111] "Version" version="v1.13.1"
>   I0515 16:20:18.787054       1 feature_gate.go:387] feature gates: {map[RecoverVolumeExpansionFailure:true]}
>   I0515 16:20:18.787221       1 envvar.go:172] "Feature gate default state" feature="ClientsAllowCBOR" enabled=false
>   I0515 16:20:18.787240       1 envvar.go:172] "Feature gate default state" feature="ClientsPreferCBOR" enabled=false
>   I0515 16:20:18.787245       1 envvar.go:172] "Feature gate default state" feature="InformerResourceVersion" enabled=false
>   I0515 16:20:18.787249       1 envvar.go:172] "Feature gate default state" feature="WatchListClient" enabled=false
>   I0515 16:20:18.788215       1 connection.go:234] "Connecting" address="unix:///csi/csi-provisioner.sock"
>   I0515 16:20:18.789799       1 common.go:143] "Probing CSI driver for readiness"
>   I0515 16:20:18.789842       1 connection.go:264] "GRPC call" method="/csi.v1.Identity/Probe" request="{}"
>   I0515 16:20:18.791060       1 connection.go:270] "GRPC response" response="{}" err=null
>   I0515 16:20:18.791083       1 connection.go:264] "GRPC call" method="/csi.v1.Identity/GetPluginInfo" request="{}"
>   I0515 16:20:18.792150       1 connection.go:270] "GRPC response" response="{\"name\":\"rbd.csi.ceph.com\",\"vendor_version\":\"canary\"}" err=null
>   I0515 16:20:18.792173       1 main.go:166] "CSI driver name" driverName="rbd.csi.ceph.com"
>   I0515 16:20:18.792183       1 connection.go:264] "GRPC call" method="/csi.v1.Identity/GetPluginCapabilities" request="{}"
>   I0515 16:20:18.792798       1 connection.go:270] "GRPC response" response="{\"capabilities\":[{\"service\":{\"type\":\"CONTROLLER_SERVICE\"}},{\"volume_expansion\":{\"type\":\"ONLINE\"}},{\"service\":{\"type\":\"VOLUME_ACCESSIBILITY_CONSTRAINTS\"}}]}" err=null
>   I0515 16:20:18.792822       1 connection.go:264] "GRPC call" method="/csi.v1.Controller/ControllerGetCapabilities" request="{}"
>   panic: runtime error: invalid memory address or nil pointer dereference
>   [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xa97ae2]
> 
>   goroutine 1 [running]:
>   github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripSingleValue({0x24249c8, 0xc0000fefc0}, {{}, 0x1c60a00?, 0x0?, 0x1c5c140?})
>   	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:71 +0x82
>   github.com/kubernetes-csi/csi-lib-utils/protosanitizer.stripValue({0x24249c8, 0xc0000fefc0}, {{}, 0x1c60a00?, 0x0?, 0xc00017a5c8?})
>   	/workspace/vendor/github.com/kubernetes-csi/csi-lib-utils/protosanitizer/protosanitizer.go:94 +0x1b8

This has been fixed with kubernetes-csi/csi-lib-utils#188. A build of the external-resizer sidecar needs **csi-lib-utils v0.21.0** or newer.

nixpanic committed:
rebase: update external-resizer sidecar to v1.13.2

The v1.13.2 build contains csi-lib-utils v0.21 which addresses a panic
when `GET_SNAPSHOT` is part of the ControllerCapabilities (unknown value
for an enum).

See-also: kubernetes-csi/csi-lib-utils#188
Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

nixpanic committed:
rebase: update external-snapshotter to v8.2.1

The v8.2.1 build contains csi-lib-utils v0.21 which addresses a panic
when `GET_SNAPSHOT` is part of the ControllerCapabilities (unknown value
for an enum).

See-also: kubernetes-csi/external-snapshotter#1278
Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

nixpanic committed:
rebase: update external-attacher to v4.8.1

The v4.8.1 build contains csi-lib-utils v0.21 which addresses a panic
when `GET_SNAPSHOT` is part of the ControllerCapabilities (unknown value
for an enum).

See-also: kubernetes-csi/external-snapshotter#1278
Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

nixpanic commented:
Yay, success! :partying_face:

@yati1998, would it help to have `Controller.GetSnapshot` merged for testing with fixes in the external-snapshotter?

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,27 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+	mgr := NewManager(cs.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	snapID := req.GetSnapshotId()
```
we need a validation to check snapID is not empty.

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -1709,3 +1709,27 @@ func (cs *ControllerServer) ControllerUnpublishVolume(
+	snapID := req.GetSnapshotId()
+	snapshot, err := mgr.GetSnapshotByID(ctx, snapID)
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "failed to find snapshot with ID %q: %s", snapID, err.Error())
```
not all errors can be considered as the NotFound, we need a check to return expected GRPC codes

nixpanic replied:
indeed, that needs more work.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1140,6 +1140,32 @@ func updateSnapshotDetails(ctx context.Context, rbdSnap *rbdSnapshot) error {
+		parent.ClusterID,
+		parent.ReservedID)
+	if err != nil {
+		return fmt.Errorf("failed to generate source volume-id: %w", err)
```
can we add the parent and rbdSnap to the error message?

nixpanic replied:
rbdSnap will be logged by the caller already, will include `parent` here

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1140,6 +1140,32 @@ func updateSnapshotDetails(ctx context.Context, rbdSnap *rbdSnapshot) error {
+		rbdSnap.conn.Creds,
+		poolID,
+		parent.Pool,
+		parent.ClusterID,
+		parent.ReservedID)
```
`getParent` doesnt set the ReservedID and also regarding clusterID is it possible to have parent and the snapshot is having different clusterID. i think its possible but it would be good to check and confirm this if its not possible.

nixpanic replied:
hmm, it feels wrong to return a volume (parent) that is not complete with all it's attributes. I"ll have a look how this can be done better.

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1140,6 +1140,32 @@ func updateSnapshotDetails(ctx context.Context, rbdSnap *rbdSnapshot) error {
+
+	poolID, err := util.GetPoolID(parent.Monitors, rbdSnap.conn.Creds, parent.Pool)
+	if err != nil {
+		return fmt.Errorf("failed to get the pool if for parent %q of snapshot %q: %w", parent, rbdSnap, err)
```
`pool if ` to `pool id`

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -1140,6 +1140,32 @@ func updateSnapshotDetails(ctx context.Context, rbdSnap *rbdSnapshot) error {
 	rbdSnap.VolSize = vol.VolSize
 
+	// get the parent to set rbdSnap.SourceVolumeID
+	parent, err := rbdSnap.getParent()
```
we can store the sourceVolumeID in the omap so that we can have 2 advantages
* when the parent is not present we still be able to set the SourceVolumeID which is expected
* SourceVolumeId is a required field in the SPEC that need to be always set
* If source is in different clusterID or something we dont need to worry about it as we dont need to check it in the backend(this might be extra).

nixpanic replied:
yeah, I wondered why the SourceVolumeId was not stored in the journal too.

nixpanic committed:
rebase: update external-provisioner to v5.3.0

The v5.3.0 build contains csi-lib-utils v0.22 which addresses a panic
when `GET_SNAPSHOT` is part of the ControllerCapabilities (unknown value
for an enum).

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.32

black-dragon74 commented:
@nixpanic, With the progress on VGS v1beta2 APIs, maybe we re-open this now?
